In [1]:
!pip install pandas

In [2]:
import pandas as pd

In [3]:
df = pd.read_csv("/content/movies.csv")

Q1. How do we load the dataset and see the first 5 rows?

In [4]:
df.head()

,id,name,year,rating,certificate,duration,genre,votes,gross_income,directors_id,directors_name,stars_id,stars_name,description
0,tt0068646,The Godfather,1972,9.2,R,175 min,"Crime, Drama",1798749,"134,966,411",nm0000338,Francis Ford Coppola,"nm0000008,nm0000199,nm0001001,nm0000473","Marlon Brando,Al Pacino,James Caan,Diane Keaton",The aging patriarch of an organized crime dyna...
1,tt5113044,Minions: The Rise of Gru,2022,7.3,PG,87 min,"Animation, Adventure, Comedy","1,220","134,966,411","nm0049633,nm1556070,nm3646390","Kyle Balda,Brad Ableson,Jonathan del Val","nm0136797,nm1853544,nm0000273,nm0378245","Steve Carell,Pierre Coffin,Alan Arkin,Taraji P...",The untold story of one twelve-year-old's drea...
2,tt7657566,Death on the Nile,2022,6.3,PG-13,127 min,"Crime, Drama, Mystery","121,063","134,966,411",nm0000110,Kenneth Branagh,"nm5290643,nm0000906,nm0000110,nm1258970","Tom Bateman,Annette Bening,Kenneth Branagh,Rus...","While on vacation on the Nile, Hercule Poirot ..."
3,tt8115900,The Bad Guys,2022,6.9,PG,100 min,"Animation, Adventure, Comedy","23,090","134,966,411",nm2010048,Pierre Perifel,"nm0005377,nm0549505,nm5377144,nm0732497","Sam Rockwell,Marc Maron,Awkwafina,Craig Robinson",Several reformed yet misunderstood criminal an...
4,tt10809742,Collision,2022,3.8,TV-MA,99 min,"Crime, Drama, Thriller",656,"134,966,411",nm1926494,Fabien Martorell,"nm1339181,nm0456810,nm12651040,nm7357127","Tessa Jubber,Langley Kirkwood,Zoey Sneedon,Bon...",Freedom always comes at a price.


In [5]:
df.tail()

,id,name,year,rating,certificate,duration,genre,votes,gross_income,directors_id,directors_name,stars_id,stars_name,description
19803,tt0459423,Kahaani Gudiya Ki...: True Story of a Woman,2008,6.2,Not Rated,112 min,"Biography, Drama",41,"44,435",nm1909606,Prabhakar Shukla,"nm0244890,nm0952232,nm1056425,nm6389513","Divya Dutta,Arif Zakaria,Rajpal Naurang Yadav,...","A tragic saga about a woman named Gudiya, whos..."
19804,tt0119621,Mark Twain's America in,1998,6.5,G,52 min,"Documentary, Biography",48,"2,281,741",nm0522844,Stephen Low,"nm0000843,nm0640297,nm4700843,nm1838044","Anne Bancroft,Dennis O'Connor,Kim Parr,LeeAnn ...",Interweaves Twain's life and times with the li...
19805,tt0168950,Let It Come Down: The Life of Paul Bowles,1998,6.9,Not Rated,75 min,"Documentary, Biography",125,"9,188",nm0047028,Jennifer Baichwal,"nm0101247,nm0123221,nm0155822,nm0159383","Paul Bowles,William S. Burroughs,Cherifa,Moham...",Add a Plot
19806,tt0267563,Hybrid,2000,7.4,Not Rated,92 min,"Documentary, Biography",146,"159,403","/name/nm0566144/,/name/nm1084921/","Monteith McCollum,Milford Beeghly",nm1084921,Milford Beeghly,The life story of Milford Beeghly: Iowa farmer...
19807,tt0116481,Hands on a Hardbody: The Documentary,1997,7.6,PG,98 min,"Documentary, Comedy, Drama",2167,"492,876",nm0082835,S.R. Bindler,"nm0021057,nm1828145,nm0185044,nm0190040","Sid Allen,Brent Baskin,Greg Cox,J.C. Crum",Twenty-four contestants compete in an enduranc...


In [6]:
df.sample(5)

,id,name,year,rating,certificate,duration,genre,votes,gross_income,directors_id,directors_name,stars_id,stars_name,description
15290,tt0080427,Below the Belt,1980,5.2,R,91 min,"Action, Comedy, Drama",206,"77,827",nm0288793,Robert Fowler,"nm0046483,nm0064980,nm0121810,nm0304000","Regina Baff,John C. Becher,Mildred Burke,James...",A New York City waitress decides to become a p...
7827,tt2321502,Uncanny,2015,6.2,Not Rated,85 min,"Drama, Horror, Sci-Fi","5,675","3,287",nm0505075,Matthew Leutwyler,"nm0916406,nm1134641,nm2189625,nm0933988","Mark Webber,David Clayton Rogers,Lucy Griffith...","The world's first ""perfect"" Artificial Intelli..."
6458,tt1555093,The Seasoning House,2012,6.1,Not Rated,90 min,"Drama, Horror, Thriller","10,295","17,092,453",nm0404931,Paul Hyett,"nm1133857,nm0675730,nm0397721,nm0910472","Rosie Day,Sean Pertwee,Kevin Howarth,Anna Walton",Sold at a brothel deep in the woods to work as...
16218,tt0038348,Beauty and the Beast,1946,7.9,Not Rated,93 min,"Drama, Fantasy, Romance",26345,"298,718","nm0168413,nm0167496","Jean Cocteau,René Clément","nm0544786,nm0196816,nm0664186,nm0314590","Jean Marais,Josette Day,Mila Parély,Nane Germon",A beautiful young woman takes her father's pla...
8981,tt0042593,In a Lonely Place,1950,7.9,Not Rated,94 min,"Drama, Film-Noir, Mystery","31,240","7,927",nm0712947,Nicholas Ray,"nm0000007,nm0002108,nm0522481,nm0717203","Humphrey Bogart,Gloria Grahame,Frank Lovejoy,C...",A potentially violent screenwriter is a murder...


Q2. What are the data types of each column?


In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19808 entries, 0 to 19807
Data columns (total 14 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   id              19808 non-null  object 
 1   name            19808 non-null  object 
 2   year            19808 non-null  int64  
 3   rating          19808 non-null  float64
 4   certificate     19808 non-null  object 
 5   duration        19808 non-null  object 
 6   genre           19808 non-null  object 
 7   votes           19808 non-null  object 
 8   gross_income    19808 non-null  object 
 9   directors_id    19808 non-null  object 
 10  directors_name  19808 non-null  object 
 11  stars_id        19808 non-null  object 
 12  stars_name      19808 non-null  object 
 13  description     19808 non-null  object 
dtypes: float64(1), int64(1), object(12)
memory usage: 2.1+ MB


In [8]:
df.describe()

,year,rating
count,19808.000000,19808.000000
mean,1995.546597,6.041973
std,22.801725,1.206831
min,1894.000000,1.100000
25%,1983.000000,5.300000
50%,2003.000000,6.200000
75%,2014.000000,6.900000
max,2022.000000,9.700000


In [9]:
df.duplicated().sum()

np.int64(0)

In [10]:
df.isna().sum()

,0
id,0
name,0
year,0
rating,0
certificate,0
duration,0
genre,0
votes,0
gross_income,0
directors_id,0



Q3. How do we clean the `duration` column? (e.g., `'175 min'` → `175`)

In [30]:
df["duration"] = df["duration"].str.replace("min","")

In [33]:
df["duration"] = df["duration"].str.replace(",","")

In [35]:
df["duration"] = df["duration"].astype(int)

In [36]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19808 entries, 0 to 19807
Data columns (total 14 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   id              19808 non-null  object 
 1   name            19808 non-null  object 
 2   year            19808 non-null  int64  
 3   rating          19808 non-null  float64
 4   certificate     19808 non-null  object 
 5   duration        19808 non-null  int64  
 6   genre           19808 non-null  object 
 7   votes           19808 non-null  object 
 8   gross_income    19808 non-null  object 
 9   directors_id    19808 non-null  object 
 10  directors_name  19808 non-null  object 
 11  stars_id        19808 non-null  object 
 12  stars_name      19808 non-null  object 
 13  description     19808 non-null  object 
dtypes: float64(1), int64(2), object(11)
memory usage: 2.1+ MB


Q4. How do we clean the `gross` column and `votes` column?

In [39]:
df["votes"] = df["votes"].str.replace(",","")

In [41]:
df["votes"] = df["votes"].astype(int)

In [45]:
df['gross_income']= df['gross_income'].str.replace(",","")

In [50]:
df['gross_income'] = df['gross_income'].str.replace("$","")

In [49]:
df['gross_income'] = df['gross_income'].str.replace("M","")

In [52]:
df['gross_income'] = df['gross_income'].astype(float)

In [53]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19808 entries, 0 to 19807
Data columns (total 14 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   id              19808 non-null  object 
 1   name            19808 non-null  object 
 2   year            19808 non-null  int64  
 3   rating          19808 non-null  float64
 4   certificate     19808 non-null  object 
 5   duration        19808 non-null  int64  
 6   genre           19808 non-null  object 
 7   votes           19808 non-null  int64  
 8   gross_income    19808 non-null  float64
 9   directors_id    19808 non-null  object 
 10  directors_name  19808 non-null  object 
 11  stars_id        19808 non-null  object 
 12  stars_name      19808 non-null  object 
 13  description     19808 non-null  object 
dtypes: float64(2), int64(3), object(9)
memory usage: 2.1+ MB


Another way to do it :

df['gross_income'] = pd.to_numeric(df['gross_income'] , errors="cohere")

Q5. How many missing values are in each column?

In [54]:
df.isnull().sum()

,0
id,0
name,0
year,0
rating,0
certificate,0
duration,0
genre,0
votes,0
gross_income,0
directors_id,0


Q7. Find the "Hidden Gems" (High Rating, Low Votes)

In [55]:
df.describe()

,year,rating,duration,votes,gross_income
count,19808.000000,19808.000000,19808.000000,1.980800e+04,1.980800e+04
mean,1995.546597,6.041973,101.777968,3.694303e+04,1.682198e+07
std,22.801725,1.206831,43.178721,1.129755e+05,4.642034e+07
min,1894.000000,1.100000,1.000000,5.000000e+00,0.000000e+00
25%,1983.000000,5.300000,90.000000,1.058000e+03,1.196010e+05
50%,2003.000000,6.200000,98.000000,4.233000e+03,1.158192e+06
75%,2014.000000,6.900000,111.000000,2.045350e+04,1.243278e+07
max,2022.000000,9.700000,5220.000000,2.574832e+06,9.366622e+08


In [62]:
df[(df['rating'] > 6.9) & (df['votes'] < 1058)]

,id,name,year,rating,certificate,duration,genre,votes,gross_income,directors_id,directors_name,stars_id,stars_name,description
3423,tt4106286,Tiny Detectives,2014,7.2,TV-14,2,"Short, Comedy, Crime",536,1501277.0,nm0124089,Andrew Bush,"nm0680983,nm0544718,nm1930273,nm7534064","Elliot Page,Kate Mara,DeMorge Brown,Jaclyn Fle...","A parody of True Detective about two short, fe..."
3502,tt0042953,Shakedown,1950,7.0,Approved,80,"Crime, Drama, Film-Noir",346,6310.0,nm0678928,Joseph Pevney,"nm0003318,nm0235627,nm0002046,nm0862937","Howard Duff,Peggy Dow,Brian Donlevy,Lawrence T...",Unscrupulous and opportunistic San Francisco n...
3522,tt0070552,High Crime,1973,7.0,R,100,"Crime, Drama, Thriller",885,1404.0,nm0144758,Enzo G. Castellari,"nm0626259,nm0721073,nm0926235,nm0090521","Franco Nero,Fernando Rey,James Whitmore,Delia ...",An Italian police inspector matches wits with ...
3594,tt0246879,Qurbani,1980,7.0,Not Rated,157,"Action, Crime, Drama",886,398420.0,nm0451211,Feroz Khan,"nm0451211,nm0451396,nm0023868,nm0451166","Feroz Khan,Vinod Khanna,Zeenat Aman,Amjad Khan",Two friends live happily until misunderstandin...
3980,tt13634096,The Bacon Hair,2020,8.5,TV-14,76,"Animation, Action, Adventure",90,852206.0,nm9762716,ObliviousHD,"nm5055725,nm10194796,nm12142163,nm0746445","Leah Ashley,Tom Aglio,Gen Sherard,Graham Rowat",Surrounded by towering walls and patrolled by ...
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
19799,tt1486548,Ahead of Time: The Extraordinary Journey of Ru...,2009,7.9,Not Rated,73,"Documentary, Biography",77,33024.0,nm0725053,Robert Richman,"nm0344295,nm4271124,nm0811750,nm4270360","Ruth Gruber,Louisa Herz Leopold,Dava Sobel,Lee...",A documentary on the remarkable life of Ruth G...
19800,tt0044917,Amazing Monsieur Fabre,1951,7.0,Approved,90,"Biography, Comedy, Drama",64,33024.0,nm0224543,Henri Diamant-Berger,"nm0294382,nm0479626,nm0709536,nm0845969","Pierre Fresnay,Elina Labourdette,André Randall...",Pierre Fresnay plays the title role of a dedic...
19801,tt1334479,"Yoo-Hoo, Mrs. Goldberg",2009,7.4,Unrated,92,"Documentary, Biography",236,1133662.0,nm0447476,Aviva Kempner,"nm0073764,nm3571106,nm1675671,nm0046029","Gertrude Berg,Lewis Berg,Sara Chase,Ruth Bader...","The story of the actress, writer, and broadcas..."
19802,tt1518255,Raw Faith,2010,8.3,Not Rated,94,"Documentary, Biography, Drama",71,1486.0,"/name/nm0927286/,/name/nm3628867/","Wm. Peter Wiedensmith,Marilyn Sewell",nm3628867,Marilyn Sewell,This surprisingly open and revealing documenta...


Q8. Find the "Overhyped" movies (Low Rating, High Gross)

In [63]:
df.describe()

,year,rating,duration,votes,gross_income
count,19808.000000,19808.000000,19808.000000,1.980800e+04,1.980800e+04
mean,1995.546597,6.041973,101.777968,3.694303e+04,1.682198e+07
std,22.801725,1.206831,43.178721,1.129755e+05,4.642034e+07
min,1894.000000,1.100000,1.000000,5.000000e+00,0.000000e+00
25%,1983.000000,5.300000,90.000000,1.058000e+03,1.196010e+05
50%,2003.000000,6.200000,98.000000,4.233000e+03,1.158192e+06
75%,2014.000000,6.900000,111.000000,2.045350e+04,1.243278e+07
max,2022.000000,9.700000,5220.000000,2.574832e+06,9.366622e+08


In [64]:
df[(df['rating'] < 5.3) & (df['gross_income'] > 12432780)]

,id,name,year,rating,certificate,duration,genre,votes,gross_income,directors_id,directors_name,stars_id,stars_name,description
4,tt10809742,Collision,2022,3.8,TV-MA,99,"Crime, Drama, Thriller",656,134966411.0,nm1926494,Fabien Martorell,"nm1339181,nm0456810,nm12651040,nm7357127","Tessa Jubber,Langley Kirkwood,Zoey Sneedon,Bon...",Freedom always comes at a price.
18,tt11656220,Midnight in the Switchgrass,2021,4.4,R,99,"Action, Crime, Thriller",6827,107928762.0,nm0256542,Randall Emmett,"nm0386472,nm1083271,nm0000246,nm0001305","Emile Hirsch,Megan Fox,Bruce Willis,Lukas Haas",An FBI agent and Florida State officer team up...
37,tt6749318,Speed Kills,2018,4.2,R,102,"Action, Crime, Drama",3381,130742922.0,nm9928566,Jodi Scurfield,"nm0000237,nm0935395,nm0261170,nm0922995","John Travolta,Katheryn Winnick,Jennifer Esposi...",Speedboat racing champion and multimillionaire...
40,tt5433138,F9: The Fast Saga,2021,5.2,PG-13,143,"Action, Crime, Thriller",125648,173202780.0,nm0510912,Justin Lin,"nm0004874,nm0735442,nm0108287,nm0879085","Vin Diesel,Michelle Rodriguez,Jordana Brewster...",Dom and the crew must take on an international...
47,tt1551614,Blowback,2022,4.4,R,93,"Action, Crime, Thriller",199,74283625.0,nm0847749,Tibor Takács,"nm1330276,nm1544217,nm6029604,nm0541909","Randy Couture,Cam Gigandet,Michele Plaia,Louis...","After planning a perfect heist, Nick is ready ..."
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
18575,tt1389139,When the Bough Breaks,2016,5.1,PG-13,107,"Drama, Thriller",6974,29703843.0,nm0143984,Jon Cassar,"nm0004820,nm0356021,nm0539082,nm0931324","Morris Chestnut,Regina Hall,Romany Malco,Micha...",A surrogate mom for a couple becomes dangerous...
18615,tt8170298,John Henry,2020,3.6,R,91,"Drama, Thriller",2368,21082165.0,nm3112158,Will Forbes,"nm0187719,nm4175452,nm0524839,nm0286010","Terry Crews,Jamila Velazquez,Ludacris,Ken Foree",When two immigrant kids on the run from his fo...
18616,tt6469960,An Imperfect Murder,2017,3.2,R,71,Thriller,1666,15156200.0,nm0864812,James Toback,"nm1092227,nm0000285,nm0001301,nm0131974","Sienna Miller,Alec Baldwin,Charles Grodin,Coll...",An actress's nightmare about killing her ex-bo...
18717,tt1859573,Pretty Obsession,2012,4.2,Not Rated,89,Thriller,193,12974636.0,nm0062341,Michael Baumgarten,"nm4386106,nm4672255,nm2373963,nm2949570","Brittany Nicole Kovler,Lydia Pritchett,Danny P...",A pretty eighteen year-old puts herself and he...


Q9. What are the unique certificates in the dataset?


In [67]:
df["certificate"].unique()

array(['R', 'PG', 'PG-13', 'TV-MA', 'Not Rated', 'Approved', 'TV-14',
       'NC-17', 'Unrated', 'G', 'Passed', 'GP', 'X', 'M/PG', 'M', 'TV-PG',
       'TV-G', '12', 'TV-13', 'TV-Y7-FV', 'MA-17', 'AO', '18', 'R-15',
       'R-12', 'R-18', 'PG-12', '6+', 'TV-Y7', 'TV-Y', 'MA-13'],
      dtype=object)

how many unique values are there?

In [68]:
df["certificate"].nunique()

31

Q10. How many movies were released each year?

In [75]:
df.groupby("year")["id"].count().sort_values(ascending = False)

,id
year,
2018,736
2017,714
2016,665
2019,657
2014,640
...,...
1898,1
1919,1
1903,1


In [76]:
df["year"].value_counts().sort_index(ascending = False)

,count
year,
2022,174
2021,470
2020,467
2019,657
2018,736
...,...
1913,3
1903,1
1902,2
